libs/libc/machine/arm/armv7-a/gnu/arch_memchr.S

```diff
@@ -66,6 +66,10 @@
  *
  ****************************************************************************/
 
+#include "libc.h"
+
+#if (LIBC_BUILD_FUNCTION == 0)
```

Comment:
#ifdef LIBC_BUILD_FUNCTION
and other similar place.
Note: LIBC_ARCH_PREVENT_xxx should prevent only the rom function, but not assembler optimized version.

Reply:
> Note: LIBC_ARCH_PREVENT_xxx should prevent only the rom function, but not assembler optimized version.

I didn't get it. Currently, it prevents any libc implementation other than NuttX's to be used. So, if `LIBC_ARCH_PREVENT_xxx` is selected, NuttX's C software version will be used. Please note that `<ARCH>_MEMCHR` selects `LIBC_ARCH_MEMCHR` automatically, so the guards in `libs/libc/machine/<arch>` files avoid both being built.

That is the intended behavior: ROM-defined and optimized versions are treated "equally" on this PR.

Do you mean to change this behavior?

Reply:
ROM optimized version may have some limitation, but the assembly optimized version doesn't. So, it is not good to forbid the user select the assembly version. For example, esp32s3 could avoid lose the userspace performance lose in protected build by linking with: https://github.com/apache/nuttx/tree/master/libs/libc/machine/xtensa

Reply:
I understand. Currently, we don't differentiate the ROM-defined implementation from the arch-specific (assembly version): both select `LIBC_ARCH_xxx` and this PR only prevent them to be used in the kernel or in the userspace. Specifically for the ESP32-S3, we need to use the ROM-defined version in the kernel (the assembly version won't work), but the ROM-defined can't be used in the userspace. So, we select `LIBC_ARCH_xxx` and `LIBC_ARCH_PREVENT_USER` to ensure that it won't build the software version in the kernel, but it'd be built for the userspace. If `<ARCH>_xxx` is selected, it will build the assembly version for the kernel and the software version for the userspace and it still wouldn't work for ESP32-S3. The opposite logic (`LIBC_ARCH_PREVENT_KERNEL`) would build the software version for the kernel.

I'd need to add a new flag to ensure that we won't build anything (software version or arch-specific assembly) to enable using the ROM defined (by the linker script) as expected.

Reply:
But, the flags already exist: LIBC_ARCH_xxx just indicate the target support the optimized implementation may come from assembly version under libc/machine or ROM version from chip. we need set e.g. CONFIG_ARMV8M_MEMCHR=y To really compile and link libc/machine's version.

Reply:
I'm sorry for the late response (I was working on something else).

Well, I understand that `<ARCH>_xxx` needs to be selected to compile and link libc/machine's version. However, the opposite is not true: there is no Kconfig for selecting using the ROM-defined functions. ROM-defined functions are set on linker and they are used when there are no other symbols that define the same function. That being said, please note that this PR does not solve our problem. Consider the following scenarios:

- Considering your suggestions about inverting the logic on libc/machine code (`#ifdef LIBC_BUILD_FUNCTION`):
1. No code would be built on flat build (software version or libc/machine);
2. On protected build, at least one of the `LIBC_ARCH_PREVENT_xxx` needs to be selected for the software version or the libc/machine being built;
3. If libc/machine code is selected (`CONFIG_ARMV8M_MEMCHR=y`, for instance), both the software version and libc/machine would be built for the other side of  `LIBC_ARCH_PREVENT_xxx` (on protected mode);

- Even after fixing the above issues, it still doesn't solve our problems: ESP32-S3 requires the ROM-defined `memcpy` to be used on flat build or in the kernel (on protected build) when Wi-Fi is enabled.
1. If we want the optimized assembly version to be available for the userspace, we'd need to ensure it's prevented in the kernel, as well as the software version;

That being said, I propose that we create a new set of rules like `LIBC_ARCH_ROM_PREVENT_xxx` and `LIBC_ARCH_MACHINE_PREVENT_xxx` to differentiate the usage. Considering the ESP32-S3, for instance, we'd select `LIBC_ARCH_ROM_PREVENT_USER` and `LIBC_ARCH_MACHINE_PREVENT_KERNEL` to make sure ROM-defined is not used in the userspace AND the libc/machine version is not being used in the kernel (at least when Wi-Fi is enabled). 

What do you think about it, @xiaoxiang781216 and @gustavonihei ?

Reply:
> I'm sorry for the late response (I was working on something else).
> 
> Well, I understand that `<ARCH>_xxx` needs to be selected to compile and link libc/machine's version. However, the opposite is not true: there is no Kconfig for selecting using the ROM-defined functions. ROM-defined functions are set on linker and they are used when there are no other symbols that define the same function.
> That being said, please note that this PR does not solve our problem. Consider the following scenarios:
> 
> * Considering your suggestions about inverting the logic on libc/machine code (`#ifdef LIBC_BUILD_FUNCTION`):
> 
> 1. No code would be built on flat build (software version or libc/machine);
> 2. On protected build, at least one of the `LIBC_ARCH_PREVENT_xxx` needs to be selected for the software version or the libc/machine being built;
> 3. If libc/machine code is selected (`CONFIG_ARMV8M_MEMCHR=y`, for instance), both the software version and libc/machine would be built for the other side of  `LIBC_ARCH_PREVENT_xxx` (on protected mode);
> 
> * Even after fixing the above issues, it still doesn't solve our problems: ESP32-S3 requires the ROM-defined `memcpy` to be used on flat build or in the kernel (on protected build) when Wi-Fi is enabled.
> 
> 1. If we want the optimized assembly version to be available for the userspace, we'd need to ensure it's prevented in the kernel, as well as the software version;
> 
> That being said, I propose that we create a new set of rules like `LIBC_ARCH_ROM_PREVENT_xxx` and `LIBC_ARCH_MACHINE_PREVENT_xxx` to differentiate the usage. Considering the ESP32-S3, for instance, we'd select `LIBC_ARCH_ROM_PREVENT_USER` and `LIBC_ARCH_MACHINE_PREVENT_KERNEL` to make sure ROM-defined is not used in the userspace AND the libc/machine version is not being used in the kernel (at least when Wi-Fi is enabled).
> 
> What do you think about it, @xiaoxiang781216 and @gustavonihei ?

look like workable, but it's enough to add something like LIBC_PREVENT_KERNEL and LIBC_PREVENT_USER which mean disable libc version(either c or assemble). Here is some possible combination:

1. default use libc version for both user and kernel
2. enable assemble version(e.g. CONFIG_ARMV8M_MEMCHR=y) for user and kernel
3. CONFIG_LIBC_PREVENT_KERNEL c version for user and rom for kernel
4. CONFIG_ARMV8M_MEMCHR and CONFIG_LIBC_PREVENT_KERNEL, assemble version for user and rom for kernel

Reply:
It's ready, @xiaoxiang781216 ! The solution you proposed fits perfectly. Could you please take a look?